### PR TITLE
Feat: Add live updating messages feature (#22)

### LINKED ISSUE #22: Add live-updating messages

### LINKED ISSUE #50: chore: remove a redundant get operator that catches global undefined routes
## Description
[This `get` operation](https://github.com/matthewsong18/csc337-final-project/blob/dev/backend/app.js#L15) is redundant because `indexRouter` has catched all undefined routes with [its get](https://github.com/matthewsong18/csc337-final-project/blob/dev/backend/routes/indexRouter.js#L21)

## Proposed Solution
Delete that get operation in `app.js` file from line 15-17

### PR DESCRIPTION
**Pull request by BaoNguyen09**
## Tests
- [test: Add test for checking timestamp format](https://github.com/matthewsong18/csc337-final-project/commit/6d458af0ef29c71fa1d1ef5e5a0574f625dd55cf)
- [test: add test for load_message_buffer function](https://github.com/matthewsong18/csc337-final-project/commit/789b11c0d1f824dcce389b88b2a3b9b607809113)
- [test: add test for load_poll_buffer function](https://github.com/matthewsong18/csc337-final-project/commit/51081ed83c00d6f8ac27b5e4409630a39c0bdde9)
- [test: add test for empty poll id array](https://github.com/matthewsong18/csc337-final-project/commit/56ced9d37f324340ec41e6448bc315872a2d969e)
- [test: add test for sort_by_timestamp function](https://github.com/matthewsong18/csc337-final-project/commit/cb83607184273103d603c2cf94db444840f2de53)
-  [test: add test for load_chat function](https://github.com/matthewsong18/csc337-final-project/commit/e9c6c7ee5d657108b381e207a00defe40d66a978)
- [test: add test for SSE controller](https://github.com/matthewsong18/csc337-final-project/commit/087235afba838eab728c4673b311482d11227743)
## Feat
- [feat: implement a function to validate timestampt format](https://github.com/matthewsong18/csc337-final-project/commit/3166d65064f3bc22fd5819f267c89c0c1dcc1f27)
- [feat: implement function to load message buffer](https://github.com/matthewsong18/csc337-final-project/commit/4e272832776fb2e682c6b0906a88f1bdf793a99f)
- [feat: implement a function to load poll buffer](https://github.com/matthewsong18/csc337-final-project/commit/4d5472a3277dbf83f9ec90e0087f7dc1184de944)
- [feat: handle empty poll id array](https://github.com/matthewsong18/csc337-final-project/commit/3879657ad6a63e27992f41f93ee58387ed0907fa)
- [feat: implement a function to sort messages and polls by timestamp](https://github.com/matthewsong18/csc337-final-project/commit/8304ce2cae6f6eb7fa23ca4833842b3dd7f3213e)
- [feat: add a function that can load chat buffer given timestamp, chat …](https://github.com/matthewsong18/csc337-final-project/commit/1c29e5f3f3f3a1addc151d0dfc7776da345ac4c5) 
- [feat: add function to initiate and maintain SSE](https://github.com/matthewsong18/csc337-final-project/commit/1c728a01647f5d92cb144384337ddc2a205cbdb6)
- [feat: add a route for establishing SSE connections](https://github.com/matthewsong18/csc337-final-project/commit/94d716a0a8b042c570bce50e19181ff5cfdc282d)
- [feat: add function to update clients about new message](https://github.com/matthewsong18/csc337-final-project/commit/cc7f09011ca5525c43fc225290d2d0cc63ee306b) (no tests yet)
### Fixes

- [fix: put every code into the try-catch block](https://github.com/matthewsong18/csc337-final-project/commit/6a1fcec9b8ec50f0a535912e5b9534dadd6789d3) (just sort_by_timestamp function)
- [fix: wrap code of load messages and polls buffer in try-catch block](https://github.com/matthewsong18/csc337-final-project/commit/d30200a30941e99cc001ee797ac82ba964d74bdb)
- [fix: modify message+poll schema and change error message of load_chat](https://github.com/matthewsong18/csc337-final-project/commit/be1c11fcea1d4a24b6c4a9b323866cb1821fef72)
### Builds
- [build: add uuid dependency](https://github.com/matthewsong18/csc337-final-project/commit/933c44e251a8af95d1bb8741d00bbeef086e9744)
- [build: add axios to fake http request for testing](https://github.com/matthewsong18/csc337-final-project/commit/ec49bcb94e81aab19dcabe1c881beb51c59cd963)

### Refactors + Chores
- [refactor: put callback of message post into chatController](https://github.com/matthewsong18/csc337-final-project/commit/d30cea6b067fb4d5d37ad6e3cf565e8c712bb7d2)
- [chore: remove redundant and incorrect get request](https://github.com/matthewsong18/csc337-final-project/commit/95e932a85dff887d96a92f680e506afe2769de4b)

### CODE REVIEW

File: backend/app.js
Comment (matthewsong18) [inline backend/app.js:1]: If you were getting rid of this, why not just delete it?

File: backend/controllers/__test__/chatController.test.js [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+    expect(timestamp_string).toBe(false);
+    expect(timestamp_wrong_order).toBe(false);
+    expect(timestamp_extra_characters).toBe(false);
+    expect(timestamp_empty_string).toBe(false);
+  });
+
+  // check load message buffer
+  it("should return an array of objects of messages sorted by timestamp, buffer size", async () => {
+    let message_ids = [];
+    const user_1 = await User.create({ user_name: "Alice", has_account: true });
```
Comment (matthewsong18) [inline backend/controllers/__test__/chatController.test.js]: You are never supposed to use `User.create()`. You imported UserService so I don't know why you didn't use it.

File: backend/models/Message.js [hunk 2]
```diff
@@ -3,7 +3,7 @@ const { Schema } = mongoose;
 
 const MessageSchema = new Schema({
   author: { type: Schema.Types.ObjectId, ref: "User", required: true },
-  chat: { type: Schema.Types.ObjectId, ref: "Chat", required: true },
+  chat: { type: Schema.Types.ObjectId, ref: "Chat", required: false },
```
Comment (matthewsong18) [inline backend/models/Message.js:6]: Why did you leave the required tag?

File: backend/package.json [hunk 3]
```diff
@@ -10,8 +10,10 @@
   "license": "ISC",
   "description": "",
   "dependencies": {
+    "axios": "^1.7.9",
```
Comment (matthewsong18) [inline backend/package.json:13]: Why is this not a dev dependency?

File: backend/controllers/__test__/chatController.test.js [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+    expect(is_valid_timestamp(poll_1.createdAt)).toBeTruthy();
+    
+    expect(poll_2.title).toBeDefined();
+    expect(poll_2.title).toBe("Poll Title 2");
+    expect(poll_2.options).toBeDefined();
+    expect(poll_2.options[0]._id == poll_option_ids[4].toString()).toBe(true);
+    expect(poll_2.users_voted).toBeDefined();
+    expect(poll_2.users_voted.length).toBe(3);
+    expect(poll_2.createdAt).toBeDefined();
+    expect(is_valid_timestamp(poll_2.createdAt)).toBeTruthy();
```
Comment (matthewsong18) [inline backend/controllers/__test__/chatController.test.js]: Why are you testing every little bit about the polls? All you needed was the title if you're checking order?

File: backend/controllers/__test__/chatController.test.js [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+  // check sorting polls and messages by timestamp
+  it("should return an array of messages and polls sorted in ascending time order", async () => {
+    const user = await User.create({ user_name: "Test User", has_account: true });
+    const message_ids = [];
+    const poll_ids = [];
+    // Create messages with different timestamps
+    const message1 = await Message.create({
+        author: user._id,
+        content: "Message 1",
+        createdAt: new Date("2024-12-08T10:00:00Z"),
```
Comment (matthewsong18) [inline backend/controllers/__test__/chatController.test.js:168]: You shouldn't ever be assigning createdAt yourself??? It does it automatically?

File: backend/controllers/chatController.js [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        })
+    } catch(err) {
+        throw new Error(`${err.message}`);
+    }
+}
+
+async function load_chat(chat_id, timestamp, buffer_size=10) {
+    try {
+        // validate chat_id (2 stesps)
+        const is_valid_id = mongoose.Types.ObjectId.isValid(chat_id);
```
Comment (matthewsong18) [inline backend/controllers/chatController.js]: Why do you need to test for an invalid id? The Chat will just say it doesn't exist.

File: backend/controllers/chatController.js [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        }
+
+        // validate timestamp format
+        if (!is_valid_timestamp(timestamp)) {
+            throw new Error("Timestamp requested is invalid");
+        } 
+        // validate timestamp logic
+        if (new Date(timestamp) > Date.now()) {
+            throw new Error("Timestamp requested is in the future");
+        }
```
Comment (matthewsong18) [inline backend/controllers/chatController.js]: Separate these into another function. They're just bloating up the load_chat function and making it hard to figure out what load_chat is doing.

File: backend/controllers/chatController.js [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+function is_valid_timestamp(timestamp) {
+    if (timestamp === null) {
+        return false;
+    }
+    if (typeof timestamp === "Date") {
+        return !isNaN(timestamp.getTime());
+    }
+    const date = new Date(timestamp);
+    return !isNaN(date.getTime()) && !isNaN(date.getMinutes());
+}
```
Comment (matthewsong18) [inline backend/controllers/chatController.js]: Move this under load chat since it's the first function called. Or move it under the verification function if you make it.

File: backend/controllers/__test__/chatController.test.js [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+    }
+    const messages = await load_message_buffer(message_ids, 8, Date.now());
+    // validate all fields desire
+    let message_1 = messages[0];
+    expect(message_1.author).toBeDefined();
+    expect(message_1.author._id == user_1._id.toString()).toBe(true);
+    expect(message_1.content).toBeDefined();
+    expect(message_1.content).toBe("Message #1");
+    expect(message_1.createdAt).toBeDefined();
+    expect(is_valid_timestamp(message_1.createdAt)).toBeTruthy();
```
Comment (matthewsong18) [inline backend/controllers/__test__/chatController.test.js]: Why are you testing every case? You should just expect message to work.

**Review by matthewsong18 — CHANGES_REQUESTED** [review]

File: backend/controllers/__test__/chatController.test.js [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+  
+  beforeEach(async () => {
+    await User.deleteMany({});
+    await Chat.deleteMany({});
+    await Message.deleteMany({});
+    await Poll.deleteMany({});
+    await PollOption.deleteMany({})
+  });
+
+  // Validate timestampt format
```
Comment (Claire-Vu) [inline backend/controllers/__test__/chatController.test.js]:
```suggestion
  // Validate timestamp format
```

File: backend/controllers/__test__/chatController.test.js [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+    });
+    poll_ids.push(poll_1._id);
+    const poll_2 = await Poll.create({
+        title: "Poll Title 2",
+        options: poll_option_ids.slice(4),
+        users_voted: [user_1, user_2, user_3],
+    });
+    poll_ids.push(poll_2._id);
+    
+    const polls = await load_poll_buffer(poll_ids, 8, Date.now());
```
Comment (Claire-Vu) [inline backend/controllers/__test__/chatController.test.js]: You create a polls list but I don't believe you tested it, except for checking the length. The following code only test the poll document you created previously

File: backend/controllers/chatController.js [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+            response: res
+        });
+
+        req.on("close", () => {
+            // Remove the client from the chat_id's connections
+            connections[chat_id] = connections[chat_id].filter(client => client.id !== clientId);
+            res.end(); // close the connection
+        })
+    } catch(err) {
+        throw new Error(`${err.message}`);
```
Comment (Claire-Vu) [inline backend/controllers/chatController.js]: Instead of throwing an error, send a response with error status code and a json message that describes the error

File: backend/controllers/chatController.js [hunk 13]
```diff
@@ -0,0 +1,154 @@
+const mongoose = require("mongoose");
+const path = require('path');
+const { v4: uuidv4 } = require('uuid');
+const { Chat, Message, Poll  } = require("../models/index");
+
+let connections = {};
+
+async function post_message(req, res) {
```
Comment (Claire-Vu) [inline backend/controllers/chatController.js]: is this function left unfinished on purpose so that it could be implemented later on?

File: backend/controllers/chatController.js [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+            res.end(); // close the connection
+        })
+    } catch(err) {
+        throw new Error(`${err.message}`);
+    }
+}
+
+async function load_chat(chat_id, timestamp, buffer_size=10) {
+    try {
+        // validate chat_id (2 stesps)
```
Comment (Claire-Vu) [inline backend/controllers/chatController.js]:
```suggestion
        // validate chat_id (2 steps)
```

File: backend/controllers/chatController.js [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+        // validate timestamp format
+        if (!is_valid_timestamp(timestamp)) {
+            throw new Error("Timestamp requested is invalid");
+        } 
+        // validate timestamp logic
+        if (new Date(timestamp) > Date.now()) {
+            throw new Error("Timestamp requested is in the future");
+        }
+
+        const chat = await Chat.findById({ _id: chat_id });
```
Comment (Claire-Vu) [inline backend/controllers/chatController.js]: This doesn't follow MVC conventions, the model should be the one handling all queries related to the database. It works though, so if we wanna roll with it we can keep it this way.

File: backend/controllers/chatController.js [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+
+async function load_message_buffer(message_ids, buffer_size, timestamp) {
+    try {
+        let messages = [];
+        if (message_ids.length === 0) return messages;
+        // return x messages based on timestamps provided
+        messages = await Message.find({
+            _id: { $in: message_ids },
+            createdAt: { $lte: new Date(timestamp) }, // Filter messages before the given timestamp
+        })
```
Comment (Claire-Vu) [inline backend/controllers/chatController.js]: again, should be in messageServices

File: backend/controllers/chatController.js [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+async function load_poll_buffer(poll_ids, buffer_size, timestamp) {
+    try {
+        let polls = [];
+        if (poll_ids.length === 0) return polls;
+        // return x polls based on timestamps provided
+        polls = await Poll.find({
+            _id: { $in: poll_ids},
+            createdAt: { $lte: new Date(timestamp) }, // Filter polls before the given timestamp
```
Comment (Claire-Vu) [inline backend/controllers/chatController.js]: pollServices

File: backend/controllers/chatController.js [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+        } 
+        // validate timestamp logic
+        if (new Date(timestamp) > Date.now()) {
+            throw new Error("Timestamp requested is in the future");
+        }
+
+        const chat = await Chat.findById({ _id: chat_id });
+        const messages = await load_message_buffer(chat.message, buffer_size, timestamp);
+        const polls = await load_poll_buffer(chat.polls, buffer_size, timestamp);
+        return sort_by_timestamp(messages, polls, buffer_size);
```
Comment (Claire-Vu) [inline backend/controllers/chatController.js]: Either we need to return the messages and polls as a json here so that view can format it, or we need to create another function to return it.

File: backend/models/Message.js (hunk 2 above)
Comment (Claire-Vu) [reply backend/models/Message.js:6]: I think he left it because we would need to change all our previous test cases if we changed it, we probably need to make a separate issue for this.

Comment (matthewsong18) [reply backend/models/Message.js:6]: > I think he left it because we would need to change all our previous test cases if we changed it, we probably need to make a separate issue for this.

Makes sense

File: backend/controllers/chatController.js (hunk 12 above)
Comment (matthewsong18) [reply backend/controllers/chatController.js]: Reference the code I used in [user_signup](https://github.com/matthewsong18/csc337-final-project/blob/e47155aceeb84063efb3d1a307401e8df02db521/backend/controllers/userController.js#L12-L16)

File: backend/controllers/chatController.js (hunk 13 above)
Comment (BaoNguyen09) [reply backend/controllers/chatController.js]: This one belongs to issue #20

File: backend/controllers/chatController.js (hunk 15 above)
Comment (BaoNguyen09) [reply backend/controllers/chatController.js]: So I should create ChatService.js file and put these code inside there?

File: backend/controllers/chatController.js (hunk 18 above)
Comment (BaoNguyen09) [reply backend/controllers/chatController.js]: I return an array so that we can keep the order of messages and polls in ascending order (latest -> oldest)

File: backend/controllers/__test__/chatController.test.js (hunk 1 above)
Comment (BaoNguyen09) [reply backend/controllers/__test__/chatController.test.js]: You're right.

File: backend/controllers/__test__/chatController.test.js (hunk 9 above)
Comment (BaoNguyen09) [reply backend/controllers/__test__/chatController.test.js]: I validate since i was unsure if the returned object has all those properties

File: backend/controllers/__test__/chatController.test.js (hunk 11 above)
Comment (BaoNguyen09) [reply backend/controllers/__test__/chatController.test.js]: You're right, I wrongly assumed that polls will contain the correct data so i just check the length

File: backend/controllers/__test__/chatController.test.js (hunk 4 above)
Comment (BaoNguyen09) [reply backend/controllers/__test__/chatController.test.js]: You're right, I can simplify these tests by checking the returned polls with the polls I created using their id

File: backend/controllers/__test__/chatController.test.js (hunk 5 above)
Comment (BaoNguyen09) [reply backend/controllers/__test__/chatController.test.js:168]: I want to explicitly put there so I can test if the sorting works

File: backend/controllers/chatController.js (hunk 12 above)
Comment (BaoNguyen09) [reply backend/controllers/chatController.js]: Sure

File: backend/controllers/chatController.js (hunk 15 above)
Comment (Claire-Vu) [reply backend/controllers/chatController.js]: that would adhere to the MVC architecture, but honestly because we don't have much time I'm fine with just leaving it as is

File: backend/controllers/chatController.js (hunk 6 above)
Comment (BaoNguyen09) [reply backend/controllers/chatController.js]: Here's my thought: If we put an invalid chat_id to findById, it will throw a `MongoException: Invalid object ID` which we also have to catch, so I just catch it early with a custom Error but i think this should goes into ChatService as well

File: backend/controllers/chatController.js (hunk 7 above)
Comment (BaoNguyen09) [reply backend/controllers/chatController.js]: Will make another commit to refactor all these functions following this [structure](https://gist.github.com/BaoNguyen09/c95ef1f8e79106621d31e49222b73645)

File: backend/controllers/chatController.js (hunk 8 above)
Comment (BaoNguyen09) [reply backend/controllers/chatController.js]: > Will make another commit to refactor all these functions following this [structure](https://gist.github.com/BaoNguyen09/c95ef1f8e79106621d31e49222b73645)

File: backend/package.json (hunk 3 above)
Comment (BaoNguyen09) [reply backend/package.json:13]: You're right since it's just used for testing

File: backend/controllers/chatController.js (hunk 18 above)
Comment (Claire-Vu) [reply backend/controllers/chatController.js]: yeah I get that, but we need a function that would send a valid response status code and the items in a json format so that the front end can use the information. Try doing res.status(200).send(sort_by_timestamp(messages, polls, buffer_size)) and check if the json is something that the frontend will be able to use to make the chat page

File: backend/controllers/chatController.js (hunk 7 above)
Comment (matthewsong18) [reply backend/controllers/chatController.js]: > Will make another commit to refactor all these functions following this [structure](https://gist.github.com/BaoNguyen09/c95ef1f8e79106621d31e49222b73645)

I like the updated structure.

File: backend/controllers/chatController.js (hunk 13 above)
Comment (matthewsong18) [reply backend/controllers/chatController.js]: > is this function left unfinished on purpose so that it could be implemented later on?

Yup. I'm working on this in issue #20 right now.

**Comment by matthewsong18** [comment]
@BaoNguyen09, I agree with your refactor plans. Regarding the code that should be in their relevant models, like chat_service and message_service, leave them be. Whoever does those issues should come back and fix this code. **Just focus on replacing improper user calls with `user_service`.**

File: backend/controllers/chatController.js (hunk 18 above)
Comment (BaoNguyen09) [reply backend/controllers/chatController.js]: So If we use JSON.parse() on the client-side, we can convert that string into an array to use: [reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#examples)

File: backend/package.json (hunk 3 above)
Comment (BaoNguyen09) [reply backend/package.json:13]: I made an [issue](https://github.com/matthewsong18/csc337-final-project/issues/65) for this since I don't want to mess up the packages

File: backend/controllers/chatController.js
Comment (matthewsong18) [inline backend/controllers/chatController.js:1]: Awesome. This is really easy to follow now.

**Review by matthewsong18 — APPROVED** [review]
Ok. It looks clean now and all your tests pass.
